### DISCUSSION TITLE
A possible alternative approach to raw string literals

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#5434] [Language Ideas]
I wonder if anyone has considered an alternative way of implementing this, namely introduce a new compiler (preprocessor) directive?

I think a way would be to consider this (I did do a proof of concept but not for C#)

```cs
#pushdelimiter |||
```

perhaps leading to something like:

```cs
#pushdelimiter |||
public string DoSomeStuff()
{
   string text = |||This is some text and "this is too" ""as is this even""|||;
}
#popdelimiter
```

Such a design could neatly support nesting, perhaps leading to something like:

```cs
#pushdelimiter |||
public string DoSomeStuff()
{
   string text = |||This is some text and "this is too" ""as is this even""|||;
}

#pushdelimiter """
public string SpecialText()
{
   string text = """This is also a delimited string containing this "" and this too ||| ! """;
}

#popdelimiter   // from here on the delimiter is ||| again.

#popdelimiter   // from here on the delimiter is " again.
```

The design is such that only one delimiter is active at any time during the compilation of the source file, but they are stacked.

This has advantages over my initial experiments (which was just an experiment, proof of concept and did not use preprocessor directives):

- No fixed delimiter character sequence
- Multiple nested delimeters
- Arbitrarily complex delimters

The only unusual aspect to this is the way the parser and scanner interact, where we allow the parser to dynamically add/remove tokens recognized by the scanner and stack the delimiter definitions, that was interesting to code. (In fact with the preprocessor directives there's probably no need at all for the parser to ever see any of this, the scanner itself recognizes preprocessor directives and can simply update its token table dynamically).

**Comment by CyrusNajmabadi** [comment]
In general, this makes the lexer parsing sensitive.  Right now the C# language is designed so these are distinctly independent layers.  Lexing happens the same way no matter what.

**Reply by The-Futurist** [reply]
> In general, this makes the lexer parsing sensitive. Right now the C# language is designed so these are distinctly independent layers. Lexing happens the same way no matter what.

I have no idea how it is actually implemented in the Roslyn code, but it's rather simple the scanner has definitions of tokens as sequences of characters, sometimes just a single character like PERIOD or SEMICOLON or QUOTE.

If the table maps a " to QUOTE then that table can be adjusted to instead map """ to QUOTE or ||| or ---- or whatever, providing there's no conflict with the already defined language tokens.

I imagine that the parser would not be impacted, it simply sees a stream of tokens, the characters that comprise the tokens are largely irrelevant to the parser, at least in my experience.

Granted there might be some consequences for the IDE, I can't comment on that.

**Reply by CyrusNajmabadi** [reply]
> If the table maps a " to QUOTE then that table

There is no table in roslyn's impl.  HOwever, this is also not about an impl thing.  This is about how the language is defined.  Parsing does not affect scanning, or preprocessing.  And scannign does not affect preprocessor.

> Granted there might be some consequences for the IDE

This is not about the IDE.  This is about needing to update the lang so that pp directives now can bleed out and affect scanning (esp. around what is/isn't a token), which then affects parsing.  This is far outside of what we've ever allowed the preprocessor to do.

Second, as mentioned below, this would not be a positive.  It would be complexity in language and impl to surface a more complex language feature.  Complexity in this space would be an anti-feature.  We explicitly considered it and rejected it, opting for a much simpler feature which accomplishes the same task but in a much easier to understand fashion when reading and writing code.

**Reply by The-Futurist** [reply]
> > If the table maps a " to QUOTE then that table
> 
> There is no table in roslyn's impl. HOwever, this is also not about an impl thing. This is about how the language is defined. Parsing does not affect scanning, or preprocessing. And scannign does not affect preprocessor.
> 
> > Granted there might be some consequences for the IDE
> 
> This is not about the IDE. This is about needing to update the lang so that pp directives now can bleed out and affect scanning (esp. around what is/isn't a token), which then affects parsing. This is far outside of what we've ever allowed the preprocessor to do.
> 
> Second, as mentioned below, this would not be a positive. It would be complexity in language and impl to surface a more complex language feature. Complexity in this space would be an anti-feature. We explicitly considered it and rejected it, opting for a much simpler feature which accomplishes the same task but in a much easier to understand fashion when reading and writing code.

Regarding "table" that was merely a metaphor for a general mapping, there _will be_ a mapping using characters and state, be it implemented as a `switch/case` structure or lookup table, that mapping could - in principle - be altered by a preprocessing directive (my own experiments show this can be straightforward).

I also never said "it" was "about" implementation nor did I say it was "about" the IDE, I **did** say "Granted there might be some consequences for the IDE, I can't comment on that" which is true, there likely would be some impact on that area since syntax highlighting needs to "understand" token structure and if that changes due to directives then it would need to changed so as to maintain awareness of those changes.

Regarding "complexity" if you look at my example (here it is again with some additional changes) I don't share your view that it is "complex" at all:

```cs
#delimiter.push |||
public string DoSomeStuff()
{
   return 
   |||
   This is some text and "this is too" ""as is this even""
   |||;
}

#delimiter.push """
public string SpecialText()
{
   return """This is also a delimited string containing this "" and this too ||| ! """;
}

#delimiter.pop    // from here on the delimiter is ||| again.

#delimiter.pop    // from here on the delimiter is " again.
```

So leaving implementation questions aside, I am at a loss to understand what is complex about this?

Your concern too, about the interaction between preprocessor and parser is likely moot, after all the directive `#nullable <option>` leads to just such an interaction it seems.

So leaving implementation aside, focusing on the example source code, what specifically do you find to be complex? (that there is no such concept today is not really an argument that therefore we should never entertain such a concept).

Clearly you view this disfavorably so perhaps we can simply agree to disagree at this time.

Thank you for your comments.

**Reply by HaloFour** [reply]
> ```cs
> #delimiter.pop    // from here on the delimiter is " again.
> ```

This suggests that you're changing the delimiter for all strings, not just for "raw" strings?

**Reply by 333fred** [reply]
> Your concern too, about the interaction between preprocessor and parser is likely moot, after all the directive #nullable <option> leads to just such an interaction it seems.

Nullable does not affect the parser at all: we always parse as if nullable is enabled, then later to back and warn about nullable annotations in non-nullable-enabled code. This proposal would add a new interaction between the preprocessor and scanner that I don't think we're interested in trying to make.

--------

In general, I will say that using preprocessor directives for this does not feel like a particularly C# approach. We use directives for modes, like nullable/TFM-specific code, not for controlling syntax rules. This is more than just a concern about implementation: it's a very different idiom for C# as a whole. I much, much, much prefer the variable quoting system that has been proposed currently, as it's more in line with the broader C# language.

**Reply by CyrusNajmabadi** [reply]
> So leaving implementation questions aside, I am at a loss to understand what is complex about this?

It's more complex because it requires users to understand how the preprocessor will affect scanning and parsing. 

It's more complex as now you may have to deal with virtually any char sequence being a delimiter.

**Reply by CyrusNajmabadi** [reply]
> after all the directive #nullable <option> leads to just such an interaction it seems.

As Fred mentions, it does not. The preprocessor does not change the syntax of our tokens. Or what syntax is allowed by the parser.

**Reply by CyrusNajmabadi** [reply]
> Clearly you view this disfavorably so perhaps we can simply agree to disagree at this time.

You brought up this possible alternative. I'm giving you my opinion of it...  Isn't that what you're asking for?

**Reply by CyrusNajmabadi** [reply]
> I don't share your view that it is "complex" at all:

That is much more complex for me to mentally parse and understand the meaning of. It also doesn't feel anything like c#. I'll presume the impl is cheap or free. Even with that, I don't think this would be good for the language due to how confusing that is to try to read/understand.

**Reply by The-Futurist** [reply]
> > ```cs
> > #delimiter.pop    // from here on the delimiter is " again.
> > ```
> 
> This suggests that you're changing the delimiter for all strings, not just for "raw" strings?

Yes, it does look that way and that's undesirable I think, I'm still reviewing what this can and can't do though.

The idea is to have the `#delimit` directives used sparingly, perhaps even just before and after where we define a string literal.

This is taken from a unit test of the experimental scanner:

```cs
        public const string DelimiterTest =
            @"
            #delimiter |||

            |||lkfhjsdkfhsdkfghsdkgfhdfkg|||

            ""jhdkjhdkajshdkajshdkkhsdhd""

            ";
    }
```

If this was a text file rather than C# string declaration, the latter would simply be `"jhdkjhdkajshdkajshdkkhsdhd"`.

In that test the code is not fully fleshed out so the `"` chars are still legitimate and the tokenizer "see" two literal strings (the delimiters are not part of the resulting lexeme either) - there are two ways to write a string, in fact they can coexist too since the scanner FSM once its seen `|||` does not care what is present until after the closing `|||`.

So this too passes the unit test:

![image](https://user-images.githubusercontent.com/12262952/142030287-205fe861-e652-4e77-953c-da37986c1484.png)

In each case all text between the delimiters is present in the returned token.

So I suppose there's no need after all to "stack" as I suggested before, each set of delimiters can coexist, in other words introducing another delimiter set does not impact any already defined delimiters but over doing that might be unwieldy.

Your (earlier) [proposal ](https://github.com/dotnet/csharplang/discussions/89)has this example:

```cs
string xml = @("<?xml version="1.0"?>
<root>
    <item attr="foo" />
</root>");
```

So this has me wondering of an additional form of the directive might also have merit:

```cs
#delimiter (" ") // two delimiters in this example

string xml = @("<?xml version="1.0"?>
<root>
    <item attr="foo" />
</root>");
```

This could easily be done and relies on the `@` to avoid conflicts, one could then do:

```cs
#delimiter [" "] // two delimiters in this example

string xml = @["<?xml version="1.0"?>
<root>
    <item attr="foo" />
</root>"];
```
 and so on, in these examples the #delimiter is only ever applied to `@` strings which is even neater (the experiments I've done don't have a concept of verbatim strings).

I guess that C# verbatim strings today are just a special case of a more general set of possibilities:

```cs
#delimiter " " // two delimiters in this example, replicates C# normal verbatim strings:

string path = @"C:\myfolder\myfile.txt";
```

This leads to the following being equally possible (though possibly distasteful to a purist)

```cs
#delimiter ' ' // two delimiters in this example, in this case we use apostrophes

string path = @'Yes, when he spoke he said "Good evening"';
```

**Reply by CyrusNajmabadi** [reply]
> The idea is to have the #delimit directives used sparingly, perhaps even just before and after where we define a string literal.

This means that for string literals instead of writing just a simple delimiter like `"""`, you now need the preprocessor directives.  That's a bunch of extra syntax and complexity which supports no more (or maybe one narrow) scenarios than the one brought up in the actual proposal.  That's not a good thing here.

**Reply by CyrusNajmabadi** [reply]
> Your (earlier) proposal has this example:

My proposal never had that :).  That may have been something that someone thought of.  But we explicitly rejected that (mentioned a few times now) as a negative approach to this problem space.

**Reply by HaloFour** [reply]
@CyrusNajmabadi 

> My proposal never had that :)

He was talking to me. 😉 

My original proposal in this space was modeled somewhat after C++ raw strings using a custom identifier within the delimiter.

**Reply by The-Futurist** [reply]
> > The idea is to have the #delimit directives used sparingly, perhaps even just before and after where we define a string literal.
> 
> This means that for string literals instead of writing just a simple delimiter like `"""`, you now need the preprocessor directives. That's a bunch of extra syntax and complexity which supports no more (or maybe one narrow) scenarios than the one brought up in the actual proposal. That's not a good thing here.

Consider some of the examples in the proposal, here we embed the literal example _itself_ into a string:

```
// surround the example with four ", at the front and the end:

var string cyrus_example = """"var xml = """<summary><element attr="content"/></summary>""""""";

```

and

```
// Here's an example using the above example, we must surround this example with five " chars.

var example = """""var string cyrus_example = """"var xml = """<summary><element attr="content"/></summary>"""""""""""";

```

In cases like this it is not very obvious at all where the string content actually ends and the delimiter begins, being able to define a custom delimiter leads to more readable code:

```
#delimiter (' ')

var string cyrus_example = ('var xml = """<summary><element attr="content"/></summary>"""');

```

Here the literal text is easily seen to be:

`var xml = """<summary><element attr="content"/></summary>"""`

and then consider:

```
#delimiter [' ']

var example = ['var string cyrus_example = ('var xml = """<summary><element attr="content"/></summary>"""')'];

```
Here the literal text is again easily seen to be:

 `var string cyrus_example = ('var xml = """<summary><element attr="content"/></summary>"""')`

In both of my examples there is no risk of confusion as to what characters are content and what are string delimiters.

The syntax of the `#delimiter` is just an example and could be changed, it was just chosen to get the concept clearly explained, I suppose we could have this:

```
#delimiter [' ']      // This activates the string delimiters
// code needing the delimiters goes here...
#delimiter           // this (no characters after the directive) deactivates the delimiters.
```

**Reply by 333fred** [reply]
> In cases like this it is not very obvious at all where the string content actually ends and the delimiter begins, being able to define a custom delimiter leads to more readable code:

I have 2 opinions on that case:
1. No one will write this as a single line.
2. Unless you're a compiler developer writing the test case for this code.

In general, while corner cases like this are important to consider, I think it's important to make sure that you weight them appropriately to the extra cost of trying to solve them. And I do not think this case is important enough to take a radically different approach to this problem that doesn't mesh well with C# as a language to this point.

**Reply by HaloFour** [reply]
I think the use of a preprocessor directive to alter the parsing of the language here is a non-starter, whether or not you think the results are more readable.  The team has also already considered and rejected the use of custom delimiters, so I don't know what else there is to discuss in this space.

**Reply by The-Futurist** [reply]
@HaloFour 

Well just to be _very_ clear none of the above is visible _in any way_ to the parser. The tokenizer itself recognizes preprocessor directives and in the case of this particular directive it obtains the delimiter chars and then analyzes those to dynamically add entries to the tokenizer FSM table allowing it to then recognize the delimited strings, the added entries are easily removed again too if another suitable preprocessor directive is encountered.

The parser has no idea about any of this, it is 100% confined to lexical analysis, the parser just sees string literal tokens with no idea of how they may have been delimited.

This is how I implemented this anyway, I can see there's no application of the idea to C# after discussing it so we'll leave it at that.

**Reply by CyrusNajmabadi** [reply]
> The parser has no idea about any of this

 "parsing of the language" refers to both the syntactic and lexical side of things here.  (some people call them lexical parsers after all).  Halo is correct here.  It's a non-starter to need to have preprocessor directives for something so basic.  It effectively poisons the feature by making something simple need to come with such heavyweight ceremony.

**Reply by CyrusNajmabadi** [reply]
> Consider some of the examples in the proposal, here we embed the literal example itself into a string:

That's a core use case we considered.  And it would be written clearly as:

```c#
var cyrus_example = """"
    var xml = """<summary><element attr="content"/></summary>"""
    """";
```

or like so if you prefer this alignment:

```c#
var cyrus_example = """"
                    var xml = """<summary><element attr="content"/></summary>"""
                    """";
```

This is very readable with everything you need to know to understand this literal self-contained at this point in a uniform fashion that applies to all code you may read in all contexts.

**Reply by CyrusNajmabadi** [reply]
> I can see there's no application of the idea to C#

At the very least, there needs to be something provided to address core questions about scenarios this would need to handle.  For example, (as asked below) how does your proposal deal with interpolated raw string literals?

**Reply by CyrusNajmabadi** [reply]
> In both of my examples there is no risk of confusion as to what characters are content and what are string delimiters.

This is a value judgement on your part.  I find it confusing, esp. the need to now have to know about the idea of a transient delimiter stack that may be defined at some arbitrary point above using some arbitrary syntax.  Per project, per file, and even per region of file this might be different, leading to having to keep an awareness of that at all time just to comprehend one of hte most basic constructs we have (a string literal).

As stated, we considered arbitrary delimiters and actively found them less desirable (regardless of if they came in through the preprocessor, or through some other syntactic mechanism).  I get that you are are saying you think of them as equally as good (or better), but i don't know what to tell you at this point.  That's your opinion.  It was considered.  And it was not found convincing for our goals here. :-/

**Reply by The-Futurist** [reply]
> > The parser has no idea about any of this
> 
> "parsing of the language" refers to both the syntactic and lexical side of things here. (some people call them lexical parsers after all). **Halo is correct here**. It's a non-starter to need to have preprocessor directives for something so basic. It effectively poisons the feature by making something simple need to come with such heavyweight ceremony.

@CyrusNajmabadi, there is no "correct" when we are discussing matters of taste and opinion, by all means disagree with me but please do not imply that when we disagree you are inevitably right and I am inevitably wrong. 

Why must you use emotive terms like "poisons" and "ceremony" and "non-starter"? all of these are superlatives, disparaging terms that have no place in rational polite discourse.

You write "I don't know what to tell you at this point" in which case say nothing and let the matter rest please.

I just said above I'm happy to let the matter rest but clearly you are not.

**Reply by CyrusNajmabadi** [reply]
> @CyrusNajmabadi, there is no "correct" when we are discussing matters of taste and opinion

He's correct here in calling this 'parsing'.  Parsing refers both to syntactic parsing and lexical parsing in this context.

**Reply by The-Futurist** [reply]
> > @CyrusNajmabadi, there is no "correct" when we are discussing matters of taste and opinion
> 
> He's correct here in calling this 'parsing'. Parsing refers both to syntactic parsing and lexical parsing in this context.

He is incorrect but what good will it do either of us to perpetuate this? please, once again let the matter rest.

**Reply by CyrusNajmabadi** [reply]
> Why must you use emotive terms like "poisons" and "ceremony" and "non-starter"?

Because they're part of language design.  It is a subjective process that weighs in things like this.  It is not sufficient for a solution to simply be technically capable of solving a problem.  If htat were the case, then design would be easy.  We'd just pick anything that worked.  Rather, we have to evaluate, *subjectively* what we should do.

This was a part of the previous discussion you started here as well.  In the context of C#, we evaluate ideas not simply on technical feasibility, but on subjective things like would they be a positive or a negative for the language and the goals we have in these spaces.  It feels like you want to have these discussions without that being part and parcel of how we do design discussions here.

**Reply by CyrusNajmabadi** [reply]
> He is incorrect but what good will it do either of us to perpetuate this?

He is correct.  As a term of art his using of 'parsing' is appropriate.  

> please, once again let the matter rest.

If you keep stating he is incorrect in this regard, I will continue pointing out why he is not :)

**Reply by CyrusNajmabadi** [reply]
> please, once again let the matter rest.

If you keep stating he is incorrect in this regard, I will continue pointing out why he is not :)

**Reply by The-Futurist** [reply]
> > please, once again let the matter rest.
> 
> If you keep stating he is incorrect in this regard, I will continue pointing out why he is not :)

@CyrusNajmabadi, this is not amusing nor is this a debating forum, if you _want_ to debate the matter I'd be delighted to do so, but this is not the forum and it is not equitable since you have authority here that I do not, if you want to publicly debate me on level terms then just say so, I can arrange that.

**Reply by CyrusNajmabadi** [reply]
@Korporal Perhaps this will help:

If you are asking: could you build a language and compiler stack that employed the strategy you are laying out here?

If so, the answer is: sure.  

Languages (and compiler stacks) can be whatever you want them to be.  If you want your own language to operate in this fashion, you certainly could.

--

However, if you're asking: would this proposal fit into the C# language?

Then you will ultimately have to deal with the opinions of those who use the language or help develop it as to what is a *good* or *bad* idea for the language.  For any given change, there are a literal infinite manner of ways the change could come about.  The subjective process here is entirely about whittling that down into the actual change that is felt to be appropriate for the product here.

**Reply by CyrusNajmabadi** [reply]
> nor is this a debating forum,

Discussions are exactly that.  It's a location to put for ideas and debate the value of htem.  We actually intend for discussions here to be entirely for this purpose, allowing enough understanding to be covered that we can then distill out actual proposals which can go through the LDM design process.  

> and it is not equitable

I cannot change that i work on the product and we have structured github to give privileges in that regard.  We engage here to help you get a sense for how these ideas will be received.  But as part of that it is true that you are discussing things with some people who have contributor permissions and some who do not.  That will be true when engaging with any person who works on a team/repo like this.

> I can arrange that.

You're welcome to go that route if that's what you prefer :)

**Reply by The-Futurist** [reply]
> > nor is this a debating forum,
> 
> Discussions are exactly that. It's a location to put for ideas and debate the value of htem. We actually intend for discussions here to be entirely for this purpose, allowing enough understanding to be covered that we can then distill out actual proposals which can go through the LDM design process.
> 
> > and it is not equitable
> 
> I cannot change that i work on the product and we have structured github to give privileges in that regard. We engage here to help you get a sense for how these ideas will be received. But as part of that it is true that you are discussing things with some people who have contributor permissions and some who do not. That will be true when engaging with any person who works on a team/repo like this.
> 
> > I can arrange that.
> 
> You're welcome to go that route if that's what you prefer :)

A "discussion" where one party has privileges not granted to the other is **not** a debate Cyrus therefore this is _not_ a debating forum insofar as disagreements between you and I are concerned. If you are _both_ a participant _and_ a moderator then the moderator's neutrality and impartiality is brought into question. If you are only willing to debate someone in situations where you are a moderator then that speaks volumes to me.

Many - not all - of your arguments are _technically_ fallacious, for example you often use phrases like "we on the language team felt" or "the team discussed this and we feel that". Now those assertions may well be true (but might not be) but they are examples of what is know in debating and philosophy as the [argument from authority](https://en.wikipedia.org/wiki/Argument_from_authority) i.e.

![image](https://user-images.githubusercontent.com/12262952/142087086-c8d1b030-0d7c-483d-a6d6-d2a3ab350a03.png)

Such arguments do not establish the truth or falsity of a claim (i.e. my suggestion about how to perform string delimiting is a "non-starter"), this is not to question the competence or expertise of the authority, but it does question the use of that authority to establish objective right and wrong, absolute true and false and you often use this argument.

So once again lets let this rest but please do not pretend this a debating forum, if it were you would confine your comments to moderation and leave you personal opinions and tastes out of the discussion - that is what a moderator does in a debate - at least in the real world.

**Reply by HaloFour** [reply]
The LDM is the authority, and given most of language design is subjective their "opinions" and "feelings" on the subject matter a great deal.  You invoke such argument yourself, such as suggesting that the raw string proposal accepted by the LDM is less clear and more complicated than your own.

I'm not sure what you're expecting to get out of this forum but I would suggest that it's not going to adhere to this strict framework of engagement that you seem to think that it should.

**Reply by CyrusNajmabadi** [reply]
Literally the purpose of this repo is about the authorities here engaging with those without it.  This is true for all repos.  There are people who are the owners/maintainers, and those who are not.  I'm sorry that you don't like that aspect of the nature of projects and how they can engage with communities.

**Reply by CyrusNajmabadi** [reply]
> If you are only willing to debate someone in situations where you are a moderator then that speaks volumes to me.

I literally said: you are welcome to go that route

:)

I was encouraging you here.  You seem to have taken issue with me supporting your idea.  I'm not sure how anything i could do would be seen as acceptable by you.  If you didn't like your idea, you could complain.  And even when i do you're not happy.  So what would you prefer i do with such a suggestion?  Just ignore it?  That personally feels rude to me given that you were proposing an idea and i thought you would likely not liek if i ignored it.

**Reply by CyrusNajmabadi** [reply]
> Such arguments do not establish the truth or falsity of a claim

The purpose of this repo is not to establish truth or falsity of a claim.  The purpose of this repo is to allow a public facing medium for the csharp design group to do their work and engage with the community.  This includes making use of many aspects of github's technologies for purposes that we see fit to use.  One of those pieces of technologies is 'discussions' a place where people can propose ideas and bounce them off other community and team members, debating and discussing the varying pros/cons involved.

We have decided on this and *explicitly* stated it here: https://github.com/dotnet/csharplang#discussions

![image](https://user-images.githubusercontent.com/4564579/142091654-94c94fae-7b85-447c-894f-e1f16274af43.png)

Note that we mention *"Debate"* as the very first word here and the *reason* for discussions.  We will continue to use discussions for these purposes, and that will absolutely involve subjective assessments (as debates will do) from teh parties involved.

**Comment by CyrusNajmabadi** [comment]
> Arbitrarily complex delimters

This was considered and is not considered a positive.  Arbitrary complexity is not virtuous here versus a constrained approach that still solves the problem.  Arbitrary approaches mean no consistency in codebases and more complexity around people mentally parsing things.

**Comment by CyrusNajmabadi** [comment]
@Korporal how does your proposal deal with raw interpolated strings?
